### PR TITLE
RV32 Support for SM/MPRV

### PR DESCRIPTION
**Pull request by ghost**
- RV32 support for MPRV
- RV32 option for CMAKE

### CODE REVIEW

File: sm/mprv.h [hunk 1]
```diff
@@ -3,36 +3,50 @@
...
 int copy1_from_sm(uintptr_t dst, const uint8_t *src);
-int copy8_from_sm(uintptr_t dst, const uint64_t *src);
-int copy64_from_sm(uintptr_t dst, const uint64_t *src);
+int copy_word_from_sm(uintptr_t dst, const uint64_t *src);
+int copy_block_from_sm(uintptr_t dst, const uint64_t *src);
 
 int copy1_to_sm(uint8_t *dst, uintptr_t src);
-int copy8_to_sm(uint64_t *dst, uintptr_t src);
-int copy64_to_sm(uint64_t *dst, uintptr_t src);
+int copy_word_to_sm(uint64_t *dst, uintptr_t src);
```
Comment (archshift) [inline sm/mprv.h]: I think we should change these then to take `size_t` or `uintptr_t` pointers instead, so the user is clear on what the right types are.

File: sm/mprv.h [hunk 2]
```diff
@@ -3,36 +3,50 @@
...
+int copy_block_to_sm(uint64_t *dst, uintptr_t src);
```
Comment (archshift) [inline sm/mprv.h]: just so `block` is clear here, can we define `typedef struct { size_t words[8] } mprv_block_t` and have this function take a `mprv_block_t *`?

**Review by archshift — CHANGES_REQUESTED** [review]
Other than the strong typing stuff LGTM!

File: sm/enclave.c [hunk 3]
```diff
@@ -241,7 +241,7 @@ static enclave_ret_code copy_word_to_host(uintptr_t dest_ptr, uintptr_t value)
 {
   enclave_ret_code ret = ENCLAVE_REGION_OVERLAPS;
 
-  int err = copy8_from_sm(dest_ptr, (uint64_t *)&value);
```
Comment (archshift) [inline sm/enclave.c:244]: Is this sized correctly on RV32?

Comment (ghost) [reply sm/enclave.c:244]: Yes, int-sizes are defined by XLEN, so this should be sized correctly.

File: sm/mprv.h (hunk 2 above)
Comment (ghost) [reply sm/mprv.h]: good idea!

File: sm/mprv.h [hunk 4]
```diff
@@ -2,37 +2,52 @@
 #include <stdint.h>
 #include <stdbool.h>
 
+typedef struct { size_t words[8]; } mprv_block_t;
```
Comment (dkohlbre) [inline sm/mprv.h]: change to mprv_block or similar, _t is a reserved naming convention.

Comment (archshift) [reply sm/mprv.h]: Ah, didn’t know it was reserved. Good call!

Comment (ghost) [reply sm/mprv.h]: Changed to mprv_block.

**Pull request merged**